dotCMS/src/main/java/com/dotmarketing/portlets/htmlpageasset/business/render/HTMLPageAssetRenderedAPIImpl.java

```diff
@@ -232,35 +245,43 @@ public String getPageHtml(
                 throws DotSecurityException, DotDataException {
 
         final Host host = this.hostWebAPI.getCurrentHost(request, context.getUser());
-        final IHTMLPage page = getHtmlPageAsset(context, host, request).getHTMLPage();
+        final HTMLPageUrl htmlPageUrl = getHtmlPageAsset(context, host, request);
+        final IHTMLPage page = htmlPageUrl.getHTMLPage();
 
         return new HTMLPageAssetRenderedBuilder()
                 .setHtmlPageAsset(page)
                 .setUser(context.getUser())
                 .setRequest(request)
                 .setResponse(response)
                 .setSite(host)
-                .getPageHTML();
+                .setURLMapper(htmlPageUrl.pageUrlMapper)
+                .setLive(htmlPageUrl.hasLive())
+                .getPageHTML(context.getPageMode());
     }
 
-    private HTMLPageUrl getHtmlPageAsset(
-            final PageContext context,
-            final Host host,
-            final HttpServletRequest request)
-                throws DotDataException, DotSecurityException {
+    private HTMLPageUrl getHtmlPageAsset(final PageContext context, final Host host, final HttpServletRequest request)
+            throws DotDataException, DotSecurityException {
 
         HTMLPageUrl htmlPageUrl = null;
-        IHTMLPage htmlPageAsset = findPageByContext(host, context);
+        for (final SearchPageFunction pageSearcher : pageSearchers) {
+            final Optional<HTMLPageUrl> optional = pageSearcher.search(context, host, request);
 
-        if (htmlPageAsset == null){
-            htmlPageUrl   = findByURLMap(context, host, request);
-            htmlPageAsset = getPageByUri(context.getPageMode(), host, htmlPageUrl.getPageUrl());
-        } else {
-            htmlPageUrl   = new HTMLPageUrl(htmlPageAsset);
+            if (optional.isPresent()) {
+                htmlPageUrl = optional.get();
+                break;
+            }
+        }
+
+        if(htmlPageUrl == null ){
+            throw new HTMLPageAssetNotFoundException(context.getPageUri());
         }
 
-        htmlPageUrl.setHTMLPage((HTMLPageAsset) htmlPageAsset);
+        checkPagePermission(context, htmlPageUrl.htmlPage);
 
+        return htmlPageUrl;
+    }
+
+    private void checkPagePermission(PageContext context, IHTMLPage htmlPageAsset) throws DotDataException, DotSecurityException {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Parameter 'htmlPageAsset' is not assigned and could be declared final](https://app.codacy.com/manual/dotCMS/core/pullRequest?prid=4842460)